File: 01_projects/code/typescript-playwright/cbaps-dawms/cbaps/pages/RequisitionPage.ts
/**
 * RequisitionPage: Comprehensive requisition creation and management
 * Enhanced with validation, getters, and utility methods
 */

import { Page, Locator } from '@playwright/test';
import { PlaywrightManager } from '../library/PlaywrightManager';
import { RequisitionData, RequisitionStatus, FundType, ValidationResult } from '../library/types';
import { FundingLinesPage } from './FundingLinesPage';
import { RoutingApprovalPage } from './RoutingApprovalPage';

export class RequisitionPage {
  readonly page: Page;
  readonly pwManager: PlaywrightManager;

  // Form Locators
  readonly titleInput: Locator;
  readonly descriptionTextarea: Locator;
  readonly fundTypeDropdown: Locator;
  readonly priorityDropdown: Locator;
  readonly submitButton: Locator;
  readonly saveAsDraftButton: Locator;
  readonly cancelButton: Locator;

  // Status and Navigation Locators
  readonly statusBadge: Locator;
  readonly requisitionIdLabel: Locator;
  readonly goToFundingLink: Locator;
  readonly routeForApprovalButton: Locator;
  readonly editButton: Locator;
  readonly deleteButton: Locator;

  // Validation Locators
  readonly titleError: Locator;
  readonly fundTypeError: Locator;
  readonly formErrorsPanel: Locator;
  readonly successMessage: Locator;

  // Info Locators
  readonly createdByLabel: Locator;
  readonly createdDateLabel: Locator;
  readonly lastModifiedLabel: Locator;

  constructor(page: Page, pwManager: PlaywrightManager) {
    this.page = page;
    this.pwManager = pwManager;

    // Initialize form locators
    this.titleInput = page.locator('#requisitionTitle');
    this.descriptionTextarea = page.locator('#requisitionDescription');
    this.fundTypeDropdown = page.locator('#fundType');
    this.priorityDropdown = page.locator('#priority');
    this.submitButton = page.locator("button:has-text('Submit')");
    this.saveAsDraftButton = page.locator("button:has-text('Save as Draft')");
    this.cancelButton = page.locator("button:has-text('Cancel')");

    // Initialize status and navigation locators
    this.statusBadge = page.locator('#reqStatus');
    this.requisitionIdLabel = page.locator('#requisitionId');
    this.goToFundingLink = page.locator("a:has-text('Funding Lines')");
    this.routeForApprovalButton = page.locator("button:has-text('Route for Approval')");
    this.editButton = page.locator("button:has-text('Edit')");
    this.deleteButton = page.locator("button:has-text('Delete')");

    // Initialize validation locators
    this.titleError = page.locator('#titleError');
    this.fundTypeError = page.locator('#fundTypeError');
    this.formErrorsPanel = page.locator('.form-errors');
    this.successMessage = page.locator('.success-message');

    // Initialize info locators
    this.createdByLabel = page.locator('#createdBy');
    this.createdDateLabel = page.locator('#createdDate');
    this.lastModifiedLabel = page.locator('#lastModified');

    // Stability anchor
    this.pwManager.waitVisible('#requisitionTitle', 'Requisition page loaded');
    console.log('✓ Requisition Page initialized');
  }

  /**
   * Create a new requisition with comprehensive data
   */
  async createRequisition(data: RequisitionData): Promise<void> {
    console.log('📝 Creating requisition with data:', data);

    // Enter title
    await this.pwManager.type(this.titleInput, data.title);

    // Enter description if provided
    if (data.description) {
      await this.pwManager.type(this.descriptionTextarea, data.description);
    }

    // Select fund type
    await this.pwManager.selectDropdown(this.fundTypeDropdown, data.fundType);

    // Select priority if provided
    if (data.priority) {
      await this.pwManager.selectDropdown(this.priorityDropdown, data.priority);
    }

    // Submit the form
    await this.pwManager.click(this.submitButton);
    console.log('✓ Requisition created successfully');

    // Wait for success message
    await this.waitForSuccessMessage();
  }

  /**
   * Fill requisition form without submitting
   */
  async fillRequisitionForm(data: RequisitionData): Promise<void> {
    await this.pwManager.type(this.titleInput, data.title);
    
    if (data.description) {
      await this.pwManager.type(this.descriptionTextarea, data.description);
    }
    
    await this.pwManager.selectDropdown(this.fundTypeDropdown, data.fundType);
    
    if (data.priority) {
      await this.pwManager.selectDropdown(this.priorityDropdown, data.priority);
    }

    console.log('✓ Requisition form filled');
  }

  /**
   * Save requisition as draft
   */
  async saveAsDraft(data: RequisitionData): Promise<void> {
    await this.fillRequisitionForm(data);
    await this.pwManager.click(this.saveAsDraftButton);
    console.log('✓ Requisition saved as draft');
    await this.waitForSuccessMessage();
  }

  /**
   * Get current requisition status
   */
  async getStatus(): Promise<string> {
    const status = await this.pwManager.getText(this.statusBadge);
    console.log(`✓ Current status: ${status}`);
    return status;
  }

  /**
   * Get requisition ID
   */
  async getRequisitionId(): Promise<string> {
    if (await this.pwManager.isVisible(this.requisitionIdLabel)) {
      return await this.pwManager.getText(this.requisitionIdLabel);
    }
    return '';
  }

  /**
   * Validate status matches expected
   */
  async validateStatus(expectedStatus: RequisitionStatus): Promise<boolean> {
    const actualStatus = await this.getStatus();
    const isValid = actualStatus === expectedStatus;
    
    if (isValid) {
      console.log(`✅ Status validation passed: ${expectedStatus}`);
    } else {
      console.error(`❌ Status mismatch: Expected "${expectedStatus}", got "${actualStatus}"`);
    }
    
    return isValid;
  }

  /**
   * Navigate to Funding Lines page
   */
  async goToFundingLines(): Promise<FundingLinesPage> {
    await this.pwManager.click(this.goToFundingLink);
    console.log('✓ Navigating to Funding Lines page');
    return new FundingLinesPage(this.page, this.pwManager);
  }

  /**
   * Route requisition for approval
   */
  async routeForApproval(): Promise<RoutingApprovalPage> {
    await this.pwManager.click(this.routeForApprovalButton);
    console.log('✓ Routing requisition for approval');
    return new RoutingApprovalPage(this.page, this.pwManager);
  }

  /**
   * Validate form before submission
   */
  async validateForm(): Promise<ValidationResult> {
    const errors: string[] = [];

    // Check title
    const title = await this.titleInput.inputValue();
    if (!title || title.trim().length === 0) {
      errors.push('Title is required');
    }

    // Check fund type
    const fundType = await this.fundTypeDropdown.inputValue();
    if (!fundType) {
      errors.push('Fund type is required');
    }

    // Check for displayed validation errors
    if (await this.pwManager.isVisible(this.titleError)) {
      const errorText = await this.pwManager.getText(this.titleError);
      errors.push(`Title error: ${errorText}`);
    }

    if (await this.pwManager.isVisible(this.fundTypeError)) {
      const errorText = await this.pwManager.getText(this.fundTypeError);
      errors.push(`Fund type error: ${errorText}`);
    }

    const result: ValidationResult = {
      isValid: errors.length === 0,
      errors
    };

    if (result.isValid) {
      console.log('✅ Form validation passed');
    } else {
      console.warn('⚠️  Form validation failed:', errors);
    }

    return result;
  }

  /**
   * Check if edit button is visible/enabled
   */
  async canEdit(): Promise<boolean> {
    return await this.pwManager.isVisible(this.editButton) && 
           await this.pwManager.isEnabled(this.editButton);
  }

  /**
   * Check if delete button is visible/enabled
   */
  async canDelete(): Promise<boolean> {
    return await this.pwManager.isVisible(this.deleteButton) && 
           await this.pwManager.isEnabled(this.deleteButton);
  }

  /**
   * Get requisition metadata
   */
  async getRequisitionMetadata(): Promise<{
    id: string;
    status: string;
    createdBy?: string;
    createdDate?: string;
    lastModified?: string;
  }> {
    const metadata: any = {
      id: await this.getRequisitionId(),
      status: await this.getStatus()
    };

    if (await this.pwManager.isVisible(this.createdByLabel)) {
      metadata.createdBy = await this.pwManager.getText(this.createdByLabel);
    }

    if (await this.pwManager.isVisible(this.createdDateLabel)) {
      metadata.createdDate = await this.pwManager.getText(this.createdDateLabel);
    }

    if (await this.pwManager.isVisible(this.lastModifiedLabel)) {
      metadata.lastModified = await this.pwManager.getText(this.lastModifiedLabel);
    }

    console.log('✓ Requisition metadata retrieved:', metadata);
    return metadata;
  }

  /**
   * Wait for success message to appear
   */
  private async waitForSuccessMessage(timeout: number = 10000): Promise<void> {
    try {
      await this.page.waitForSelector('.success-message', { state: 'visible', timeout });
      console.log('✓ Success message displayed');
    } catch {
      console.warn('⚠️  Success message did not appear');
    }
  }

  /**
   * Clear form
   */
  async clearForm(): Promise<void> {
    await this.titleInput.clear();
    await this.descriptionTextarea.clear();
    console.log('✓ Form cleared');
  }

  /**
   * Cancel requisition creation
   */
  async cancel(): Promise<void> {
    await this.pwManager.click(this.cancelButton);
    console.log('✓ Requisition creation cancelled');
  }

  /**
   * Check if routing button is available
   */
  async isRoutingAvailable(): Promise<boolean> {
    return await this.pwManager.isVisible(this.routeForApprovalButton) &&
           await this.pwManager.isEnabled(this.routeForApprovalButton);
  }

  /**
   * Check if funding link is available
   */
  async isFundingAvailable(): Promise<boolean> {
    return await this.pwManager.isVisible(this.goToFundingLink);
  }
}

File: 01_projects/code/java-playwright/cbaps-dawms/robust/src/main/java/com/playwright/cbaps/pages/RequisitionPage.java
package com.playwright.cbaps.pages;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.playwright.cbaps.library.EnhancedPlaywrightManager;
import com.playwright.cbaps.models.RequisitionData;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Enhanced RequisitionPage - 22+ comprehensive methods
 * Matches TypeScript version's robustness with validations, calculations, and state checks
 */
public class RequisitionPage {
    private static final Logger log = LoggerFactory.getLogger(RequisitionPage.class);
    
    private Page page;
    private EnhancedPlaywrightManager pwm;
    
    // Locators
    private Locator titleInput;
    private Locator descriptionInput;
    private Locator fundTypeDropdown;
    private Locator priorityDropdown;
    private Locator submitButton;
    private Locator saveAsDraftButton;
    private Locator cancelButton;
    private Locator statusBadge;
    private Locator requisitionIdLabel;
    private Locator fundingLinesLink;
    private Locator routeForApprovalButton;
    private Locator lastModifiedLabel;
    private Locator createdByLabel;
    
    public RequisitionPage(Page page, EnhancedPlaywrightManager pwm) {
        this.page = page;
        this.pwm = pwm;
        
        // Initialize locators
        this.titleInput = page.locator("#requisitionTitle");
        this.descriptionInput = page.locator("#description");
        this.fundTypeDropdown = page.locator("#fundType");
        this.priorityDropdown = page.locator("#priority");
        this.submitButton = page.locator("button:has-text('Submit')");
        this.saveAsDraftButton = page.locator("button:has-text('Save as Draft')");
        this.cancelButton = page.locator("button:has-text('Cancel')");
        this.statusBadge = page.locator("#reqStatus");
        this.requisitionIdLabel = page.locator("#requisitionId");
        this.fundingLinesLink = page.locator("a:has-text('Funding Lines')");
        this.routeForApprovalButton = page.locator("button:has-text('Route for Approval')");
        this.lastModifiedLabel = page.locator("#lastModified");
        this.createdByLabel = page.locator("#createdBy");
        
        // Stability anchor - wait for page to load
        pwm.waitUntilElementVisible("#requisitionTitle");
        log.info("✅ RequisitionPage initialized");
    }
    
    // ========== Business Methods ==========
    
    /**
     * Create requisition with complete data and submit
     */
    public void createRequisition(RequisitionData data) {
        log.info("📝 Creating requisition: {}", data.getTitle());
        
        pwm.enterText(titleInput, data.getTitle());
        
        if (data.getDescription() != null && !data.getDescription().isEmpty()) {
            pwm.enterText(descriptionInput, data.getDescription());
        }
        
        pwm.selectDropdown(fundTypeDropdown, data.getFundType());
        
        if (data.getPriority() != null && !data.getPriority().isEmpty()) {
            pwm.selectDropdown(priorityDropdown, data.getPriority());
        }
        
        pwm.clickElement(submitButton);
        log.info("✅ Requisition created successfully");
    }
    
    /**
     * Fill form without submitting
     */
    public void fillRequisitionForm(RequisitionData data) {
        log.info("📋 Filling requisition form");
        
        pwm.enterText(titleInput, data.getTitle());
        
        if (data.getDescription() != null) {
            pwm.enterText(descriptionInput, data.getDescription());
        }
        
        pwm.selectDropdown(fundTypeDropdown, data.getFundType());
        
        if (data.getPriority() != null) {
            pwm.selectDropdown(priorityDropdown, data.getPriority());
        }
        
        log.info("✅ Form filled");
    }
    
    /**
     * Save requisition as draft
     */
    public void saveAsDraft(RequisitionData data) {
        log.info("💾 Saving as draft");
        fillRequisitionForm(data);
        pwm.clickElement(saveAsDraftButton);
        log.info("✅ Saved as draft");
    }
    
    /**
     * Clear all form fields
     */
    public void clearForm() {
        log.info("🧹 Clearing form");
        titleInput.fill("");
        descriptionInput.fill("");
        log.info("✅ Form cleared");
    }
    
    /**
     * Cancel requisition creation
     */
    public void cancel() {
        pwm.clickElement(cancelButton);
        log.info("❌ Requisition creation cancelled");
    }
    
    // ========== Navigation Methods ==========
    
    /**
     * Navigate to Funding Lines page
     */
    public FundingLinesPage goToFundingLines() {
        pwm.clickElement(fundingLinesLink);
        log.info("➡️ Navigating to Funding Lines");
        return new FundingLinesPage(page, pwm);
    }
    
    /**
     * Navigate to Routing/Approval page
     */
    public RoutingApprovalPage routeForApproval() {
        pwm.clickElement(routeForApprovalButton);
        log.info("➡️ Routing for approval");
        return new RoutingApprovalPage(page, pwm);
    }
    
    // ========== Validation Methods ==========
    
    /**
     * Validate form has required fields filled
     */
    public boolean validateForm() {
        boolean isValid = true;
        String title = pwm.getText(titleInput);
        
        if (title == null || title.trim().isEmpty()) {
            log.warn("⚠️ Title is required");
            isValid = false;
        }
        
        if (!pwm.isVisible(fundTypeDropdown) || pwm.getText(fundTypeDropdown).isEmpty()) {
            log.warn("⚠️ Fund type is required");
            isValid = false;
        }
        
        if (isValid) {
            log.info("✅ Form validation passed");
        } else {
            log.error("❌ Form validation failed");
        }
        
        return isValid;
    }
    
    /**
     * Validate requisition status matches expected
     */
    public boolean validateStatus(String expectedStatus) {
        String actualStatus = getStatus();
        boolean isValid = actualStatus.equals(expectedStatus);
        
        if (isValid) {
            log.info("✅ Status validation passed: {}", expectedStatus);
        } else {
            log.error("❌ Status mismatch: Expected '{}', Got '{}'", expectedStatus, actualStatus);
        }
        
        return isValid;
    }
    
    /**
     * Validate requisition ID exists
     */
    public boolean validateRequisitionId() {
        String reqId = getRequisitionId();
        boolean isValid = reqId != null && !reqId.isEmpty();
        
        if (isValid) {
            log.info("✅ Requisition ID validation passed: {}", reqId);
        } else {
            log.error("❌ Requisition ID missing");
        }
        
        return isValid;
    }
    
    // ========== Getter Methods ==========
    
    /**
     * Get current requisition status
     */
    public String getStatus() {
        String status = pwm.getText(statusBadge);
        log.info("📊 Status: {}", status);
        return status;
    }
    
    /**
     * Get requisition ID
     */
    public String getRequisitionId() {
        if (pwm.isVisible(requisitionIdLabel)) {
            String id = pwm.getText(requisitionIdLabel);
            log.info("🆔 Requisition ID: {}", id);
            return id;
        }
        return null;
    }
    
    /**
     * Get requisition metadata as formatted string
     */
    public String getRequisitionMetadata() {
        StringBuilder metadata = new StringBuilder();
        metadata.append("ID: ").append(getRequisitionId()).append(", ");
        metadata.append("Status: ").append(getStatus());
        
        if (pwm.isVisible(lastModifiedLabel)) {
            metadata.append(", Modified: ").append(pwm.getText(lastModifiedLabel));
        }
        
        if (pwm.isVisible(createdByLabel)) {
            metadata.append(", Created By: ").append(pwm.getText(createdByLabel));
        }
        
        log.info("📋 Metadata: {}", metadata.toString());
        return metadata.toString();
    }
    
    /**
     * Get requisition title
     */
    public String getTitle() {
        return pwm.getAttribute(titleInput, "value");
    }
    
    // ========== State Check Methods ==========
    
    /**
     * Check if requisition can be edited
     */
    public boolean canEdit() {
        String status = getStatus();
        boolean editable = "Draft".equalsIgnoreCase(status) || "Rejected".equalsIgnoreCase(status);
        log.info("📝 Can edit: {}", editable);
        return editable;
    }
    
    /**
     * Check if requisition can be deleted
     */
    public boolean canDelete() {
        String status = getStatus();
        boolean deletable = "Draft".equalsIgnoreCase(status);
        log.info("🗑️ Can delete: {}", deletable);
        return deletable;
    }
    
    /**
     * Check if routing is available
     */
    public boolean isRoutingAvailable() {
        boolean available = pwm.isVisible(routeForApprovalButton) && pwm.isEnabled(routeForApprovalButton);
        log.info("✍️ Routing available: {}", available);
        return available;
    }
    
    /**
     * Check if funding lines section is available
     */
    public boolean isFundingAvailable() {
        boolean available = pwm.isVisible(fundingLinesLink);
        log.info("💰 Funding available: {}", available);
        return available;
    }
    
    /**
     * Check if submit button is enabled
     */
    public boolean canSubmit() {
        return pwm.isEnabled(submitButton);
    }
    
    // ========== Utility Methods ==========
    
    /**
     * Wait for status to change to expected value
     */
    public boolean waitForStatus(String expectedStatus, int timeoutSeconds) {
        log.info("⏳ Waiting for status: {}", expectedStatus);
        
        for (int i = 0; i < timeoutSeconds; i++) {
            if (getStatus().equals(expectedStatus)) {
                log.info("✅ Status changed to: {}", expectedStatus);
                return true;
            }
            pwm.wait(1);
        }
        
        log.error("❌ Status did not change to: {}", expectedStatus);
        return false;
    }
    
    /**
     * Refresh requisition page
     */
    public void refresh() {
        page.reload();
        pwm.waitUntilElementVisible("#requisitionTitle");
        log.info("🔄 Page refreshed");
    }
}
